torch/_refs/__init__.py

```diff
@@ -1655,48 +1655,108 @@ def item(a: TensorLikeType) -> NumberType:
 def _to_will_alias(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    copy: bool,
-    memory_format: torch.memory_format,
+    device: Optional[torch.device],
+    dtype: Optional[torch.dtype],
+    copy: Optional[bool],
+    memory_format: Optional[torch.memory_format],
+    layout: Optional[torch.layout],
 ) -> bool:
     return (
         not copy
-        and a.dtype == dtype
+        and (device is None or a.device == device)
+        and (dtype is None or a.dtype == dtype)
+        and (layout is None or a.layout == layout)
         and (
-            memory_format == torch.preserve_format
-            or a.is_contiguous(memory_format=memory_format)
+            memory_format is None
+            or memory_format == torch.preserve_format
+            or utils.is_contiguous_for_memory_format(a, memory_format=memory_format)
         )
     )
 
 
-# TODO: there are multiple overloads of to, but we only support one here
-# aten/src/ATen/native/native_functions.yaml lists all overloads
-# Missing overloads:
-# to.dtype_layout
-# to.device
-# to.other
-# https://github.com/pytorch/pytorch/issues/84264
+def _to_dispatch(
+    device=None,
+    dtype=None,
+    non_blocking=None,
+    copy=None,
+    memory_format=None,
+    layout=None,
+    pin_memory=None,
+):
+    if isinstance(device, TensorLike):
+        # overload to `to.other`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        # extract tensor options
+        other = device
+        device = other.device
+        dtype = other.dtype
+        layout = other.layout
+        pin_memory = other.pin_memory
+    elif isinstance(device, torch.dtype):
+        # overload to `to.dtype`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        dtype = device
+        layout = None
+        pin_memory = None
+        device = None
+    # overload to `to.device`
+    # do nothing because signature of 'to' matches
+    # overload to `to.dtype_layout`
+    # do nothing because the `to.dtype_layout` takes all kwargs.
+    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+
+
 def to(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    non_blocking: bool = False,
-    copy: bool = False,
-    memory_format: torch.memory_format = torch.preserve_format,
+    device: Optional[torch.device] = None,
```

Comment:
We should add a comment about our `.to` parsing strategy. `to` is a tricky and kinda weird function. What about making the ref have the signature:

```
def to(a: TensorLikeType, *args, **kwargs)
```

and then inspecting the first element of `*args` (if it exists) to see if it's a device, dtype, or tensor? Then we can extract the properties, diff them against the tensor's current metadata to come back with a subset of properties that need to change, and from there call `empty_like` or short-circuit to just a `dtype` or `device` conversion as appropriate?

Reply:
^^^ good point. In terms of implementation, I think it's gonna look roughly about the same here. But the intention of dispatch would be cleaner comparing to swapping arguments in current `_to_dispatch`

Reply:
Lol, after I started writing this I realized that we'll have `args` and `kwargs` possibly overlapping. so we'll need more checks for that. :vomiting_face: